Split log entries so each entry contains a maximum of 100 datapoints

paulez opened the pull request:
*Issue #, if available: #10

*Description of changes:*

- Submit no more than 100 datapoints per log entry.
- Split log entries so each entry contains a maximum of 100 datapoints.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

jaredcnance commented on aws_embedded_metrics/constants.py:
```diff
@@ -14,3 +14,4 @@
 DEFAULT_NAMESPACE = "aws-embedded-metrics"
 MAX_DIMENSIONS = 9
 MAX_METRICS_PER_EVENT = 100
+MAX_DATAPOINTS_PER_EVENT = 100
```
Consider renaming to `MAX_DATAPOINTS_PER_METRIC`.

paulez replied:
Updated!

jaredcnance commented on aws_embedded_metrics/serializers/log_serializer.py:
```diff
@@ -53,24 +55,53 @@ def create_body() -> Dict[str, Any]:
         current_body: Dict[str, Any] = create_body()
         event_batches: List[str] = []
         num_metrics_in_current_body = 0
+        num_datapoints_in_current_body = 0
```
I'm not sure I'm following why we need to track the number of datapoints in the body. There are 2 requirements:
1. You cannot have more than 100 metric names in an EMF event
2. A given metric name cannot have more than 100 samples

These two requirements bound the number of metric datapoints in an event to 10_000.

paulez replied:
I didn't understand the spec properly, I thought the limit of 100 samples was for the body. I will correct to limit to 100 samples per metric name.

paulez commented:
I've updated the logic to limit to 100 datapoints per metric per batch.
I've also added a test to combine the 100 datapoints and 100 metrics per batch limit logic.

jaredcnance commented on aws_embedded_metrics/serializers/log_serializer.py:
```diff
@@ -50,29 +52,39 @@ def create_body() -> Dict[str, Any]:
+        current_body: Dict[str, Any] = {}
         event_batches: List[str] = []
         num_metrics_in_current_body = 0
+        missing_data = True
```
Can you add some comments about what this is and consider thinking about a more intuitive name? It's not really "missing" data, something more like data left to process.

paulez replied:
I've added comments and renamed to "remaining_data".

jaredcnance commented on tests/serializer/test_log_serializer.py:
```diff
@@ -130,6 +131,73 @@ def test_serialize_more_than_100_metrics():
+
+    # assert
+    datapoints_count = Counter()
+    for batch in results:
```
should we have an assertion on the number of expected events?

paulez replied:
I've added an extra assert on the batch count.

I've also added logic and assert to validate that we don't miss any datapoint with the slicing logic.

jaredcnance reviewed (approved):
Thanks for the PR and for your time!
